cylc/flow/dbstatecheck.py

```diff
@@ -151,36 +203,69 @@ def workflow_state_query(
                 stmt_args.append(state)
                 stmt_frags.append("status==?")
             stmt_wheres.append("(" + (" OR ").join(stmt_frags) + ")")
+
         if stmt_wheres:
-            stmt += " where " + (" AND ").join(stmt_wheres)
+            stmt += "WHERE\n    " + (" AND ").join(stmt_wheres)
+
+        if status:
+            stmt += dedent("""
+                ORDER BY
+                    submit_num
+            """)
+
+        # idx_status = 2
+        # idx_flow = 3
+
+        # TODO option to list outputs instead of status
+        # TODO if flows not specified, use the most recent for the task
 
         res = []
         for row in self.conn.execute(stmt, stmt_args):
-            if not all(v is None for v in row):
-                res.append(list(row))
+            if row[2] is None:
+                # status can be None in Cylc 7 DBs
+                continue
+            if not self.back_compat_mode:
+                flow_nums = deserialise(row[3])
+                if flow_num is not None and flow_num not in flow_nums:
+                    continue
+            res.append(list(row))
 
-        return res
+        if output:
+            # Replace res with a task-states like result,
+            # [[foo, 2032, output], [foo, 2033, output]]
+            if self.back_compat_mode:
+                # Cylc 7 DB: list of {output: message}
+                results = [
+                    [name, cycle, output]
+                    for name, cycle, outputs_str in res
+                    if output in deserialise(outputs_str).values()
+                ]
+            else:
+                # Cylc 8 DB list of [output]
+                results = []
+                for name, cycle, outputs_str, flows_str in res:
+                    flows = deserialise(flows_str)
+                    outputs = deserialise(outputs_str)
+                    if output not in outputs:
```

Comment:
I think that this is broken (I've included a manual test below, becuase it's possible that I've misunderstood how this command ought to work.

Looking at the db I think that it would be more accurate to call the field in res currently called `outputs_str`, `messages_string`. This condition is therefore only ever false if the output == message. That's fine for the standard internal messages, but probably not for most external ones.

## Example

```toml
# flow.cylc
[scheduler]
    allow implicit tasks = True
    cycle point format = %Y

[scheduling]
    initial cycle point = 2380
    [[graph]]
        R1 = foo:trigger => wonder_horse

[runtime]
    [[foo]]
        script = cylc message -- message
        [[[outputs]]]
            # If you tested with trigger = trigger this will not show
            trigger = 'message'   
```

```bash
cylc vip .
sleep 5

# This is never satisfied
cylc workflow-state icky.lynx --point 2380 --task foo --output trigger

# This is
cylc workflow-state icky.lynx --point 2380 --task foo --output message

```

Reply:
What should work there is `--trigger` but not `--message`.

This relates to (a) Ronnie's comment [above](https://github.com/cylc/cylc-flow/pull/5809#discussion_r1583193487) to remove `--message`; and (b) the pending local changes I mentioned (the reason for temporarily converting this PR back to Draft).

Post optional outputs etc. we decided that the graph trigger name is "the output", rather than (as was historically the case) the task message. So all user commands and queries should specify the trigger, not the message. Only the task job sends the message, which can be more verbose for logging purposes.

Reply:
> What should work there is `--trigger` but not `--message` (which isn't the case as you've noted).
> 
> Sorry, this was partially borked, hence the pending local changes mentioned as the reason for temporarily converting this PR back to Draft.
> 
> Post optional outputs, we decided that "the outputs" are now the graph trigger names, not the task messages (as was historically the case) . So now all user commands should specify the trigger, not the message. Only the task job sends the message, which can be more verbose for logging purposes.
> 
> My initial thought on this PR was to alias `--message` to `--output` so that existing use cases would continue to work for (the majority of?) cases where trigger and message are identical. Then as per Ronnie's comment above better to just remove `--message` altogther.
> 
> However, I'm rethinking that. It would be easy enough to continue to support both, just in this command, to avoid breaking any existing workflows that use it for triggering (NIWA has many).

This sounds reasonable, especailly if there's a deprecation warning if `--message` is used.

Reply:
> especially if there's a deprecation warning if `--message` is used.

👍

Reply:
Unfortunately we've often used `[outputs] x = x` for convenience (laziness?) in examples and tests. Which of course fails to pick up any confusion between trigger and message in the code.